go-controller/pkg/ovn/egressip.go

```diff
@@ -1945,6 +1934,91 @@ func (e *egressIPZoneController) deleteEgressIPStatusSetup(name string, status e
 	return podIPs, nil
 }
 
+func (e *egressIPZoneController) ensureOnlyValidNextHops(name string, ops []libovsdb.Operation) ([]libovsdb.Operation, error) {
+	// When no nextHopIP is found, This may happen when node object is already deleted.
+	// So compare validNextHopIPs associated with current eIP.Status and Nexthops present
+	// in the LogicalRouterPolicy, then delete nexthop(s) from LogicalRouterPolicy if
+	// it doesn't match with nexthops derived from eIP.Status.
+	policyPred := func(item *nbdb.LogicalRouterPolicy) bool {
+		return item.Priority == types.EgressIPReroutePriority && item.ExternalIDs["name"] == name
+	}
+	eIP, err := e.watchFactory.GetEgressIP(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return ops, fmt.Errorf("error retrieving EgressIP %s object for updating logical router policy nexthops, err: %w", name, err)
+	} else if err != nil && apierrors.IsNotFound(err) {
+		// EgressIP object is not found, so delete LRP associated with it.
+		ops, err = libovsdbops.DeleteLogicalRouterPolicyWithPredicateOps(e.nbClient, ops, e.GetNetworkScopedClusterRouterName(), policyPred)
+		if err != nil {
+			return ops, fmt.Errorf("error creating ops to remove logical router policy for EgressIP %s from router %s: %v",
+				name, e.GetNetworkScopedClusterRouterName(), err)
+		}
+	} else {
+		validNextHopIPs := make(sets.Set[string])
+		for _, validStatus := range eIP.Status.Items {
+			nextHopIP, err := e.attemptToGetNextHopIP(name, validStatus)
+			if err != nil {
+				return ops, fmt.Errorf("failed to delete EgressIP %s (%s) because unable to determine next hop: %v",
+					name, validStatus.EgressIP, err)
+			}
+			validNextHopIPs.Insert(nextHopIP)
+		}
+
+		reRoutePolicies, err := libovsdbops.FindLogicalRouterPoliciesWithPredicate(e.nbClient, policyPred)
+		if err != nil {
+			return ops, fmt.Errorf("error finding logical router policy for EgressIP %s: %v", name, err)
+		}
+		if len(validNextHopIPs) == 0 {
+			err = libovsdbops.DeleteLogicalRouterPolicies(e.nbClient, e.GetNetworkScopedClusterRouterName(), reRoutePolicies...)
+			if err != nil {
+				return ops, fmt.Errorf("error removing logical router policy for EgressIP %s from router %s: %v",
+					name, e.GetNetworkScopedClusterRouterName(), err)
+			}
```

Comment:
nit; add return here - looks like we are ok because `DeleteNextHopsFromLogicalRouterPolicyOps` wont fail if something is deleted but it looks weird that we could be iterating over policies that are now deleted.